sim: Update multisim to use Process instead of Pool

Harshil2107 opened the pull request:
- Using `Multiprocessing.Pool` in multisim was causing errors when number of jobs were > 4x the number pf processes as a worker thread was being used to run a gem5 simulation after already running one.

- This change updates the multisim code to manually spawn the processes and handle them.

Harshil2107 committed:
sim: Manually close the multiprocessing pool

- The last process to complete threw  `OSError: [Errno 9] Bad file descriptor` error as pool was being closed by __del__

- Manually closing the pool makes sure that all processes are complete before we close the pool.

Harshil2107 commented:
This commit fixes the error that occurs when the number of jobs are <= 4x number of processes. I am looking into the error that occurs when number of jobs > 4x number of processes.

Harshil2107 committed:
sim: Update multisim to use Process instead of Pool

Harshil2107 committed:
sim: remove print statements

Harshil2107 committed:
sim: Add better error handling in multisim

powerjg commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -172,8 +176,10 @@ def _run(module_path: Path, id: str) -> None:
-
-    sim_list[0].run()
+    try:
+        sim_list[0].run()
```
Did you test this? Try putting an error in the simulator (e.g., maybe in the exit event handler?) and see what happens.

Harshil2107 replied:
Yes, I raised an exeption in exit event handler and it was caught and the error message was printed.

```
info: Error running simulator fs-sim: Exit event handler not implemented
```

The other simulations were not affected by this. I tested by running this fullsystem simulation with other SE mode simulation. The SE mode simulation ran as expected irrespective if the FS simulation was run first or last.

powerjg commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -200,17 +206,53 @@ def run(module_path: Path, processes: Optional[int] = None) -> None:
+    active_processes = []
+    remaining_ids = list(ids).copy()
+    process_lock = Lock()
+    from gem5.utils.multiprocessing import Process
```
The old code used relative imports... Why change it?

powerjg commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -200,17 +206,53 @@ def run(module_path: Path, processes: Optional[int] = None) -> None:
+                    with process_lock:
+                        active_processes.append(process)
+                except Exception as e:
+                    inform(f"Error starting process for {id_to_run}: {e}")
```
Do we want to move on, exit, or re-raise the exception?

Harshil2107 replied:
I think we can throw the exeption for the simulator but then move on to other simulators.

powerjg commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -200,17 +206,53 @@ def run(module_path: Path, processes: Optional[int] = None) -> None:
+            # Wait for active processes to finish
+            time.sleep(1)  # Avoid busy-waiting
+            with process_lock:
+                active_processes = [
```
Why the "copy" here? I have a gut feeling why you chose to do this, but it would be good to document in a comment why you made that decision.

Harshil2107 replied:
Just to clarify, you mean why I used `.copy` to initialize the variable?

If yes that was so that the original list is not modified.

powerjg replied:
No, I mean that you use a list comprehension instead of directly modifying `active_processes`

Harshil2107 replied:
I did that, as if we did it during a while loop we are editing the list we are iterating over and that might cause errors. I will add it as a comment.

powerjg reviewed (commented):
From copilot (the second point is important to address)

### Review of Changes in `multisim.py`

The changes in the `multisim.py` file are aimed at improving the process handling for the gem5 MultiSim module. Here are the key modifications and their review:

1. **Imports and Global Variables**:
   - Added `signal`, `time`, and `Lock` from `threading`.
   - Added `inform` from `m5.util`.

2. **Error Handling in `_run` Function**:
   - Wrapped the simulator run call in a try-except block to log errors using `inform`.

3. **Process Management in `run` Function**:
   - Replaced the use of `multiprocessing.Pool` with manual process management using `Process` from `gem5.utils.multiprocessing`.
   - Introduced a lock (`process_lock`) for thread-safe modifications to the `active_processes` list.
   - Added signal handlers (`SIGINT` and `SIGTERM`) to ensure processes are cleaned up on termination.

4. **Main Process Loop**:
   - Managed starting new processes and handling active ones within a loop.
   - Added a sleep interval to avoid busy-waiting.
   - Ensured proper cleanup of processes in a finally block.

### Concerns and Suggestions

1. **Busy-Waiting**:
   - The use of `time.sleep(1)` helps mitigate busy-waiting but might need tuning based on performance needs.

2. **Thread-Safe Process List**:
   - Using `threading.Lock` is not suitable for multiprocessing. Instead, consider using `multiprocessing.Lock` or other synchronization mechanisms provided by the multiprocessing module.

3. **Signal Handling**:
   - Ensure that the signal handlers are sufficient for all environments where the code will run, as some platforms may handle signals differently.

4. **Exception Handling**:
   - The try-except blocks provide basic error logging. Consider adding more granular error handling if specific error types need different handling logic.

5. **Resource Management**:
   - Verify that all resources are properly managed and released, especially if the processes open files or network connections.

### Conclusion
The changes address the initial issues with `multiprocessing.Pool` by implementing a more controlled process management system. Ensure to test thoroughly to handle edge cases and platform-specific behaviors.

erin-le reviewed (commented):
I don't have a thorough understanding of the code itself, but I tested this PR using the multisim testing scripts that Bobby provided in the [issue](https://github.com/gem5/gem5/issues/1738) and all of the cases passed. These scripts test combinations of 1 process and 1 job up to 10 processes and 10 jobs, for a total of 100 cases.

powerjg commented:
> I don't have a thorough understanding of the code itself, but I tested this PR using the multisim testing scripts that Bobby provided in the [issue](https://github.com/gem5/gem5/issues/1738) and all of the cases passed. These scripts test combinations of 1 process and 1 job up to 10 processes and 10 jobs, for a total of 100 cases.

@erin-le did you test thinks like 10 processes and 100 jobs? 10 processes 3 jobs?

Harshil2107 commented:
I locally tested this change using Bobby's script with 10 processes and 50 jobs, It works properly but it looks like its slower in spawning processes as the jobs increase, which I assume is due to running the config script again and again for 50 simulator objects for every simulation.

Harshil2107 committed:
sim: Address requested changes

powerjg commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -51,13 +51,17 @@
+import signal
+import time
 from pathlib import Path
+from threading import Lock
```
This should be
```suggestion
from multiprocessing import Lock
```

The threading lock is not compatible with multiprocessing.

powerjg reviewed (commented):
One more comment, then this looks good to me. We should let Bobby give it a review before merging, though

Harshil2107 committed:
sim: Update Lock import

- Update import statement to import Lock from multiprocessing instead  of Threading

BobbyRBruce commented:
> * Using `Multiprocessing.Pool` in multisim was causing errors when number of jobs were > 4x the number pf processes as a worker thread was being used to run a gem5 simulation after already running one.

Can you explain a bit more what exactly `Pool` was doing to cause this bug? I don't really understand this explanation. Why did 4x cause the issues?

powerjg commented:
> Can you explain a bit more what exactly Pool was doing to cause this bug? I don't really understand this explanation. Why did 4x cause the issues?

It simply didn't work. There were many different combinations of number of processes and the number of jobs such that it didn't work. We have to override the way multiprocessing works to get things to work with the gem5 binary, and the default `multiprocessing.Pool` is incompatible with this.

Let's merge this change since it makes multisim work better.
